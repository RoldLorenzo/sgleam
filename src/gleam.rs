use camino::{Utf8Path, Utf8PathBuf};
use gleam_core::{
    build::{
        Mode, Module, NullTelemetry, PackageCompiler, StaleTracker, Target,
        TargetCodegenConfiguration, Telemetry,
    },
    config::PackageConfig,
    error::{FileIoAction, FileKind},
    io::{memory::InMemoryFileSystem, FileSystemWriter},
    javascript::PRELUDE,
    type_::ModuleFunction,
    uid::UniqueIdGenerator,
    warning::{WarningEmitter, WarningEmitterIO},
    Error, Warning,
};
use std::{
    collections::HashSet,
    io::{IsTerminal, Read, Write},
    path::PathBuf,
    rc::Rc,
    time::{Duration, Instant, SystemTime},
};
use tar::Archive;
use termcolor::{BufferWriter, Color, ColorChoice, ColorSpec, WriteColor};

pub struct Project {
    pub fs: InMemoryFileSystem,
}

impl Default for Project {
    fn default() -> Project {
        let mut project = Project {
            fs: InMemoryFileSystem::new(),
        };
        extract_tar(
            &mut project.fs,
            Archive::new(crate::GLEAM_STDLIB),
            Project::source(),
        )
        .expect("Extract gleam-stdlib.tar");
        project.write_source("sgleam/check.gleam", crate::SGLEAM_CHECK);
        project.write_source("sgleam_ffi.mjs", crate::SGLEAM_FFI_MJS);
        project.write_out("prelude.mjs", PRELUDE);
        project
    }
}

impl Project {
    pub fn root() -> &'static Utf8Path {
        "/".into()
    }

    pub fn source() -> &'static Utf8Path {
        "/src".into()
    }

    pub fn out() -> &'static Utf8Path {
        "/build".into()
    }

    pub fn prelude() -> &'static Utf8Path {
        "/build/prelude.mjs".into()
    }

    pub fn write_source(&mut self, name: &str, content: &str) {
        let path = Project::source().join(name);
        self.fs
            .write(&path, content)
            .expect("Write a file in memory");
        self.fs
            .try_set_modification_time(&path, SystemTime::now())
            .expect("Set modification time of a file in memory")
    }

    pub fn copy_to_source(&mut self, input: &Utf8Path) -> Result<(), Error> {
        let content = std::fs::read_to_string(input).map_err(|err| Error::FileIo {
            kind: FileKind::File,
            action: FileIoAction::Read,
            path: input.into(),
            err: Some(err.to_string()),
        })?;
        let to = input.file_name().expect("Extract file name");
        self.write_source(to, &content);
        Ok(())
    }

    pub fn write_out(&mut self, name: &str, content: &str) {
        let path = Project::out().join(name);
        self.fs
            .write(&path, content)
            .expect("Write a file in memory");
    }
}

pub fn show_gleam_error(err: Error) {
    let buffer_writer = stderr_buffer_writer();
    let mut buffer = buffer_writer.buffer();
    err.pretty(&mut buffer);
    buffer_writer
        .print(&buffer)
        .expect("Write warning to stderr");
}

pub fn get_module(modules: Vec<Module>, name: &str) -> Option<Module> {
    modules.into_iter().find(|m| m.name == name)
}

pub fn get_main_function(module: &Module) -> Result<ModuleFunction, Error> {
    module.ast.type_info.get_main_function(Target::JavaScript)
}

pub fn compile(project: &mut Project, repl: bool) -> Result<Vec<Module>, Error> {
    let config = PackageConfig {
        target: Target::JavaScript,
        ..Default::default()
    };

    let target = TargetCodegenConfiguration::JavaScript {
        emit_typescript_definitions: false,
        prelude_location: Project::prelude().into(),
    };

    let mut compiler = PackageCompiler::new(
        &config,
        Mode::Dev,
        Project::root(),
        Project::out(),
        Project::out(),
        &target,
        UniqueIdGenerator::new(),
        project.fs.clone(),
    );

    compiler.write_metadata = true;

    compiler
        .compile(
            &WarningEmitter::new(Rc::new(ConsoleWarningEmitter::with_repl(repl))),
            &mut im::HashMap::new(),
            &mut im::HashMap::new(),
            &mut StaleTracker::default(),
            &mut HashSet::new(),
            &NullTelemetry,
        )
        .into_result()
}

fn extract_tar(
    fs: &mut InMemoryFileSystem,
    mut arch: Archive<&[u8]>,
    to: &Utf8Path,
) -> Result<(), Error> {
    let mut buf = vec![];
    for entry in arch.entries().map_err(to_error_stdio)? {
        let mut entry = entry.map_err(to_error_stdio)?;
        let is_dir = entry.header().entry_type().is_dir();
        let entry_path = entry.path().map_err(to_error_stdio)?.into_owned();
        let entry_path = Utf8PathBuf::from_path_buf(entry_path).map_err(to_error_nonutf8_path)?;
        let path = to.join(entry_path);
        if is_dir {
            fs.mkdir(&path)?;
        } else {
            buf.clear();
            entry.read_to_end(&mut buf).map_err(to_error_stdio)?;
            fs.write_bytes(&path, &buf)?;
        }
    }
    Ok(())
}

fn to_error_stdio(err: std::io::Error) -> Error {
    Error::StandardIo {
        action: gleam_core::error::StandardIoAction::Read,
        err: Some(err.kind()),
    }
}

pub fn to_error_nonutf8_path(path: PathBuf) -> Error {
    Error::NonUtf8Path { path }
}

// The remaining of this file is copied from gleam project

#[derive(Debug, Default, Clone)]
pub struct Reporter;

impl Reporter {
    pub fn new() -> Self {
        Self
    }
}

impl Telemetry for Reporter {
    fn compiled_package(&self, duration: Duration) {
        print_compiled(duration);
    }

    fn compiling_package(&self, name: &str) {
        print_compiling(name);
    }

    fn checked_package(&self, duration: Duration) {
        print_checked(duration);
    }

    fn checking_package(&self, name: &str) {
        print_checking(name);
    }

    fn downloading_package(&self, name: &str) {
        print_downloading(name)
    }

    fn packages_downloaded(&self, start: Instant, count: usize) {
        print_packages_downloaded(start, count)
    }

    fn resolving_package_versions(&self) {
        print_resolving_versions()
    }

    fn running(&self, name: &str) {
        print_running(name);
    }

    fn waiting_for_build_directory_lock(&self) {
        print_waiting_for_build_directory_lock()
    }
}

pub fn print_published(duration: Duration) {
    print_colourful_prefix("Published", &format!("in {}", seconds(duration)))
}

pub fn print_retired(package: &str, version: &str) {
    print_colourful_prefix("Retired", &format!("{package} {version}"))
}

pub fn print_unretired(package: &str, version: &str) {
    print_colourful_prefix("Unretired", &format!("{package} {version}"))
}

pub fn print_publishing_documentation() {
    print_colourful_prefix("Publishing", "documentation");
}

fn print_downloading(text: &str) {
    print_colourful_prefix("Downloading", text)
}

fn print_waiting_for_build_directory_lock() {
    print_colourful_prefix("Waiting", "for build directory lock")
}

fn print_resolving_versions() {
    print_colourful_prefix("Resolving", "versions")
}

fn print_compiling(text: &str) {
    print_colourful_prefix("Compiling", text)
}

pub(crate) fn print_checking(text: &str) {
    print_colourful_prefix("Checking", text)
}

pub(crate) fn print_compiled(duration: Duration) {
    print_colourful_prefix("Compiled", &format!("in {}", seconds(duration)))
}

pub(crate) fn print_checked(duration: Duration) {
    print_colourful_prefix("Checked", &format!("in {}", seconds(duration)))
}

pub(crate) fn print_running(text: &str) {
    print_colourful_prefix("Running", text)
}

fn print_packages_downloaded(start: Instant, count: usize) {
    let elapsed = seconds(start.elapsed());
    let msg = match count {
        1 => format!("1 package in {elapsed}"),
        _ => format!("{count} packages in {elapsed}"),
    };
    print_colourful_prefix("Downloaded", &msg)
}

pub fn seconds(duration: Duration) -> String {
    format!("{:.2}s", duration.as_millis() as f32 / 1000.)
}

pub fn print_colourful_prefix(prefix: &str, text: &str) {
    let buffer_writer = stderr_buffer_writer();
    let mut buffer = buffer_writer.buffer();
    buffer
        .set_color(
            ColorSpec::new()
                .set_intense(true)
                .set_fg(Some(Color::Magenta)),
        )
        .expect("print_colourful_prefix");
    write!(buffer, "{prefix: >11}").expect("print_colourful_prefix");
    buffer
        .set_color(&ColorSpec::new())
        .expect("print_colourful_prefix");
    writeln!(buffer, " {text}").expect("print_colourful_prefix");
    buffer_writer
        .print(&buffer)
        .expect("print_colourful_prefix");
}

pub fn stderr_buffer_writer() -> BufferWriter {
    // Don't add color codes to the output if standard error isn't connected to a terminal
    BufferWriter::stderr(color_choice())
}

fn colour_forced() -> bool {
    if let Ok(force) = std::env::var("FORCE_COLOR") {
        !force.is_empty()
    } else {
        false
    }
}

fn color_choice() -> ColorChoice {
    if colour_forced() {
        ColorChoice::Always
    } else if std::io::stderr().is_terminal() {
        ColorChoice::Auto
    } else {
        ColorChoice::Never
    }
}

#[derive(Debug, Clone, Copy)]
pub struct ConsoleWarningEmitter {
    repl: bool,
}

impl ConsoleWarningEmitter {
    pub fn with_repl(repl: bool) -> ConsoleWarningEmitter {
        ConsoleWarningEmitter { repl }
    }
}

impl WarningEmitterIO for ConsoleWarningEmitter {
    fn emit_warning(&self, warning: Warning) {
        if self.repl {
            if let Warning::Type {
                warning:
                    gleam_core::type_::Warning::UnusedImportedValue { .. }
                    | gleam_core::type_::Warning::UnusedImportedModule { .. }
                    | gleam_core::type_::Warning::UnusedImportedModuleAlias { .. },
                ..
            } = warning
            {
                return;
            }
        }
        let buffer_writer = stderr_buffer_writer();
        let mut buffer = buffer_writer.buffer();
        warning.pretty(&mut buffer);
        buffer_writer
            .print(&buffer)
            .expect("Write warning to stderr");
    }
}
