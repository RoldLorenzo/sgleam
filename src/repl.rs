use std::{collections::HashMap, fmt::Write};

use gleam_core::{build::Module, io::FileSystemWriter, Error};
use indoc::formatdoc;
use rquickjs::{Array, Context};

use crate::{
    error::{show_error, SgleamError},
    gleam::{compile, get_module, type_to_string, Project},
    javascript::{self, MainKind},
    repl_reader::ReplReader,
    run::get_main,
    swrite, swriteln, GLEAM_MODULES_NAMES,
};

const FNS_REPL: &str = r#"
@external(javascript, "./sgleam_ffi.mjs", "repl_save")
pub fn repl_save(value: a) -> a

@external(javascript, "./sgleam_ffi.mjs", "repl_load")
pub fn repl_load(index: Int) -> a
"#;

pub const QUIT: &str = ":quit";
pub const TYPE: &str = ":type ";

pub fn welcome_message() -> String {
    format!(
        "Welcome to {}.\nType ctrl-d ou \"{QUIT}\" to exit.\n",
        crate::version()
    )
}

#[derive(Clone)]
pub struct Repl {
    user_import: Option<String>,
    imports: Vec<String>,
    consts: Vec<String>,
    types: Vec<String>,
    fns: Vec<String>,
    vars: HashMap<String, (usize, String)>,
    project: Project,
    context: Context,
    type_: bool,
    iter: usize,
    var_index: usize,
}

enum EntryKind {
    Let(String, String),
    Expr(String),
    Other,
}

impl Repl {
    pub fn new(project: Project, user_module: Option<&Module>) -> Result<Repl, SgleamError> {
        let imports = GLEAM_MODULES_NAMES.iter().map(|s| s.to_string()).collect();
        let fs = project.fs.clone();
        Ok(Repl {
            user_import: user_module.map(import_public_types_and_values),
            imports,
            consts: vec![],
            types: vec![],
            fns: vec![],
            vars: HashMap::new(),
            project,
            context: javascript::create_context(fs, Project::out().into())?,
            type_: false,
            iter: 0,
            var_index: 0,
        })
    }

    pub fn run(&mut self) -> Result<(), SgleamError> {
        let editor = ReplReader::new()?;
        for mut code in editor {
            let code_trim = code.trim();
            if code_trim.is_empty() || code_trim.starts_with("//") {
                continue;
            }

            if code_trim == QUIT {
                break;
            }

            if let Some(expr) = code_trim.strip_prefix(TYPE) {
                self.type_ = true;
                code = expr.into();
            } else {
                self.type_ = false;
            }

            self.iter += 1;

            // FIXME: avoid this clone
            // We clone self so we can rollback if the execution fail
            let mut repl = (*self).clone();

            let code_no_pub = code.trim_start().strip_prefix("pub ").unwrap_or(&code);
            let pub_code = format!("pub {code_no_pub}");
            let result = match (code_no_pub.split_whitespace().next(), self.type_) {
                (Some("import"), false) => repl.run_import(code),
                (Some("const"), false) => repl.run_const(pub_code),
                (Some("type"), false) => repl.run_type(pub_code),
                (Some("let"), false) => repl.run_let(code),
                (Some("fn"), false) => repl.run_fn(pub_code),
                _ => repl.run_expr(code),
            };

            if let Err(err) = result {
                show_error(&err.into());
            } else {
                // rollback
                *self = repl;
            }
        }
        Ok(())
    }

    fn run_code(&mut self, kind: EntryKind) -> Result<(), Error> {
        let mut src = String::new();
        src.push_str(FNS_REPL);
        self.add_imports(&mut src);
        self.add_consts(&mut src);
        self.add_types(&mut src);
        self.add_fns(&mut src);

        let ret = if self.type_ { "" } else { "Nil" };

        match &kind {
            EntryKind::Let(_, expr) => {
                // FIXME: can we generate code that generates better error messagens?
                // Examples of entries that generates poor errors
                // "pub "
                // "let"
                let lets = self.get_lets();
                src.push_str(&formatdoc! {"
                    pub fn main() {{
                    {lets}
                      io.debug(repl_save({{
                        {expr}
                      }}))
                      {ret}
                    }}
                    "
                });
            }
            EntryKind::Expr(expr) => {
                let lets = self.get_lets();
                src.push_str(&formatdoc! {"
                    pub fn main() {{
                      {lets}
                      io.debug({{
                        {expr}
                      }})
                      {ret}
                    }}
                    "
                });
            }
            _ => {
                // main function is not needed
            }
        }

        let iter = self.iter;
        let module_name = format!("repl{iter}");
        let file = format!("{module_name}.gleam");

        // TODO: add an option to show the generated code
        self.project.write_source(&file, &src);

        let result = compile(&mut self.project, true);

        if let Ok(modules) = &result {
            let module = get_module(modules, &module_name).expect("The repl module");
            if let EntryKind::Let(_, _) | EntryKind::Expr(_) = &kind {
                if self.type_ {
                    let type_ = get_main(module).expect("main function").return_type.clone();
                    println!("{}", type_to_string(type_));
                } else {
                    javascript::run_main(&self.context, MainKind::Nil, &module_name);
                }
            } else {
                // Nothing to run, was a definition (type, const, import or fn)
            }

            if let EntryKind::Let(name, _) = &kind {
                if self.try_save_var(name, self.var_index, module) {
                    self.var_index += 1;
                }
            }
        }

        self.project
            .fs
            .delete_file(&Project::source().join(file))
            .expect("To delete repl file");

        result.map(|_| ())
    }

    fn run_import(&mut self, _code: String) -> Result<(), Error> {
        println!("imports are not supported.");
        Ok(())
        // TODO: implement import merge
        // import gleam/string.{append}
        // import gleam/string.{inspect}
        // -> import gleam/string.{append, inspect}
        // let new_import = code.trim().strip_prefix("import ").unwrap_or("");
        // self.imports.push(new_import.into());
        // self.run_code(EntryKind::Other)
    }

    fn run_const(&mut self, code: String) -> Result<(), Error> {
        // TODO: improve error message for const redefinition
        self.consts.push(code);
        self.run_code(EntryKind::Other)
    }

    fn run_type(&mut self, code: String) -> Result<(), Error> {
        // TODO: improve error message for type redefinition
        self.types.push(code);
        self.run_code(EntryKind::Other)
    }

    fn run_fn(&mut self, code: String) -> Result<(), Error> {
        if let Some((pub_fn_name, code)) = code.split_once('(') {
            if let Some(name) = pub_fn_name.strip_prefix("pub fn").map(str::trim) {
                if !name.contains(char::is_whitespace) {
                    // TODO: check if the compiler erros are ok
                    return self.run_let(format!("let {name} = fn({code}"));
                }
            }
        }
        // We could not transforme the code to a let expression, so it can be an anonymous function
        self.run_code(EntryKind::Expr(code))
    }

    fn run_let(&mut self, code: String) -> Result<(), Error> {
        if let Some(name) = code
            .trim()
            .strip_prefix("let")
            .and_then(|s| s.split_once('=').map(|s| s.0))
            .map(|s| s.split_once(':').map(|s| s.0).unwrap_or(s))
            .map(str::trim)
        {
            if name.chars().all(|c| c.is_alphanumeric() || c == '_') {
                return self.run_code(EntryKind::Let(name.into(), code));
            } else {
                println!("Only let with single names are supported.");
                return Ok(());
            }
        }
        // We could not get the binding name, so we run it to fail
        self.run_code(EntryKind::Expr(code))
    }

    fn run_expr(&mut self, code: String) -> Result<(), Error> {
        self.run_code(EntryKind::Expr(code))
    }

    fn add_imports(&self, src: &mut String) {
        if let Some(user) = &self.user_import {
            swriteln!(src, "{user}");
        }
        for import in &self.imports {
            swriteln!(src, "import {import}");
        }
    }

    fn add_consts(&self, src: &mut String) {
        for const_ in &self.consts {
            swriteln!(src, "{const_}");
        }
    }

    fn add_types(&self, src: &mut String) {
        for type_ in &self.types {
            swriteln!(src, "{type_}");
        }
    }

    fn add_fns(&self, src: &mut String) {
        for fn_ in &self.fns {
            swriteln!(src, "{fn_}");
        }
    }

    fn get_lets(&mut self) -> String {
        let mut lets = String::new();
        for (name, (index, ty)) in &self.vars {
            swriteln!(lets, r#"  let {name}: {ty} = repl_load({index})"#);
        }
        lets
    }

    fn try_save_var(&mut self, name: &str, index: usize, module: &Module) -> bool {
        if !self.context.with(|ctx| {
            ctx.globals()
                .get::<_, Array>("repl_vars")
                .map(|a| index < a.len())
                .unwrap_or(false)
        }) {
            // the expression crashed and repl_save was not called
            return false;
        }

        let return_type = module
            .ast
            .definitions
            .iter()
            .filter_map(|d| d.main_function())
            .next()
            .expect("The main function")
            .return_type
            .clone();

        self.vars
            .insert(name.into(), (index, type_to_string(return_type)));

        true
    }
}

fn import_public_types_and_values(module: &Module) -> String {
    let mut import = String::new();
    let name = &module.name;
    swrite!(&mut import, "import {name}.{{");
    for type_ in module.ast.type_info.public_type_names() {
        swrite!(&mut import, "type {type_}, ");
    }
    for value in module.ast.type_info.public_value_names() {
        swrite!(&mut import, "{value}, ");
    }
    import.push('}');
    import
}
