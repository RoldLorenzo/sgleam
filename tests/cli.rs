use assert_cmd::prelude::*;
use indoc::formatdoc;
use sgleam::repl::welcome_message;

use std::{
    io::Write,
    process::{Command, Stdio},
};

#[test]
fn repl_smoke_test() {
    let lit = "13\nTrue\n\"casa\"\nOk(Nil)\n781239812731283189237890123781923";
    assert_eq!(repl_exec(lit), lit);
}

#[test]
fn repl_user_module_import() {
    let input = concat!(env!("CARGO_MANIFEST_DIR"), "/tests/inputs/user.gleam");
    assert_eq!(
        repl_exec_args(&["-q", "-i", input], "one\ntwo()\nlet _: Three = Num3"),
        "1\n2\nNum3\n"
    );
}

#[test]
fn format_stdin() {
    assert_eq!(
        repl_exec_args(
            &["-f"],
            &formatdoc! {r#"
        import gleam / io.{{ debug , }}
        fn main() {{
            debug("Hello world!" )
        }}
    "#},
        ),
        formatdoc! {r#"
        import gleam/io.{{debug}}

        fn main() {{
          debug("Hello world!")
        }}
        "#}
    )
}

#[test]
fn repl_welcome_message() {
    assert_eq!(repl_exec_args(&[], ""), welcome_message())
}

fn repl_exec(s: &str) -> String {
    repl_exec_args(&["-q"], s)
        .strip_suffix('\n')
        .unwrap()
        .into()
}

fn repl_exec_args(args: &[&str], s: &str) -> String {
    let mut cmd = Command::cargo_bin(env!("CARGO_PKG_NAME")).unwrap();

    let mut child = cmd
        .stdin(Stdio::piped())
        .stdout(Stdio::piped())
        .args(args)
        .spawn()
        .expect("Spawn child process");

    let mut stdin = child.stdin.take().expect("Open stdin");

    let s = format!("{s}\n");
    std::thread::spawn(move || {
        stdin.write_all(s.as_bytes()).expect("Write to stdin");
    });

    let output = child.wait_with_output().expect("Read stdout");
    String::from_utf8_lossy(&output.stdout).to_string()
}
