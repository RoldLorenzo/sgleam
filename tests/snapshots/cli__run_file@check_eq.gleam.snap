---
source: tests/cli.rs
expression: "formatdoc!\n{\n    \"\n            STDOUT\n            {out}\n            STDERR\n            {err}\"\n}"
input_file: tests/inputs/check_eq.gleam
---
STDOUT

STDERR
warning: Unused private function
  ┌─ /src/check_eq.gleam:3:1
  │
3 │ fn ignore_examples() {
  │ ^^^^^^^^^^^^^^^^^^^^ This private function is never used

Hint: You can safely remove it.

error: Module does not have a main function

`check_eq` does not have a main function so the module can not be run.
Hint: Add a public `main` function to to `src/check_eq.gleam`.
